src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,299 @@
+  exact continuous_map.congr_fun (set.Icc_extend_of_right_le (@zero_le_one ℝ _) F.curry ht) x,
+end
+
+@[simp]
+lemma extend_apply_coe (F : homotopy_with f₀ f₁ P) (t : I) (x : X) : F.extend t x = F (t, x) :=
+continuous_map.congr_fun (set.Icc_extend_coe (@zero_le_one ℝ _) F.curry t) x
+
+@[simp]
+lemma extend_apply_of_mem_I (F : homotopy_with f₀ f₁ P) {t : ℝ} (ht : t ∈ I) (x : X) :
+  F.extend t x = F (⟨t, ht⟩, x) :=
+continuous_map.congr_fun (set.Icc_extend_of_mem (@zero_le_one ℝ _) F.curry ht) x
+
+lemma extend_prop (F : homotopy_with f₀ f₁ P) (t : ℝ) : P (F.extend t) :=
+begin
+  by_cases ht₀ : 0 ≤ t,
+  { by_cases ht₁ : t ≤ 1,
+    { convert F.prop ⟨t, ht₀, ht₁⟩,
+      ext,
+      rw [F.extend_apply_of_mem_I ⟨ht₀, ht₁⟩, F.curry_apply] },
+    { convert F.prop 1,
+      ext,
+      rw [F.extend_apply_of_one_le (le_of_not_le ht₁), F.curry_apply, F.apply_one] } },
+  { convert F.prop 0,
+    ext,
+    rw [F.extend_apply_of_le_zero (le_of_not_le ht₀), F.curry_apply, F.apply_zero] }
+end
+
+end
+
+variable {P : C(X, Y) → Prop}
+
+/--
+Given a continuous function `f` and a proof `hP : P f`, we can define a `homotopy_with f f` by
+`F (t, x) = f x`
+-/
+def refl (f : C(X, Y)) (hP : P f) : homotopy_with f f P :=
+{ to_fun := λ x, f x.2,
+  continuous_to_fun := by continuity,
+  to_fun_zero := λ _, rfl,
+  to_fun_one := λ _, rfl,
+  prop' := λ t, begin
+    convert hP,
+    ext, refl,
+  end }
+
+@[simp]
+lemma refl_apply {f : C(X, Y)} (hP : P f) (x : I × X) : refl f hP x = f x.2 := rfl
+
+instance : inhabited (homotopy_with (continuous_map.id : C(X, X)) continuous_map.id (λ f, true)) :=
+⟨homotopy_with.refl continuous_map.id trivial⟩
+
+/--
+Given a `homotopy_with f₀ f₁ P`, we can define a `homotopy_with f₁ f₀ P` by reversing the homotopy.
+-/
+def symm {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) : homotopy_with f₁ f₀ P :=
+{ to_fun := λ x, F (σ x.1, x.2),
+  continuous_to_fun := by continuity,
+  to_fun_zero := by norm_num,
+  to_fun_one := by norm_num,
+  prop' := λ t, by simpa using F.prop' (σ t) }
+
+@[simp]
+lemma symm_apply {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (x : I × X) :
+  F.symm x = F (σ x.1, x.2) := rfl
+
+@[simp]
+lemma symm_symm {f₀ f₁ : C(X, Y)} (F : homotopy_with f₀ f₁ P) : F.symm.symm = F :=
+by { ext, simp }
+
+/--
+Given `homotopy_with f₀ f₁ P` and `homotopy_with f₁ f₂ P`, we can define a `homotopy_with f₀ f₂ P`
+by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+-/
+def trans {f₀ f₁ f₂ : C(X, Y)} (F : homotopy_with f₀ f₁ P) (G : homotopy_with f₁ f₂ P) :
+  homotopy_with f₀ f₂ P :=
+{ to_fun := λ x, if (x.1 : ℝ) ≤ 1/2 then F.extend (2 * x.1) x.2 else G.extend (2 * x.1 - 1) x.2,
+  continuous_to_fun := begin
+    refine continuous_if_le _ _ (continuous.continuous_on _) (continuous.continuous_on _) _,
+    swap 5,
+    { rintros x hx,
+      norm_num [hx] },
+    exact continuous_induced_dom.comp continuous_fst,
+    exact continuous_const,
+    -- TODO: Work out why `continuity` doesn't work here.
```

Comment:
The problem is that `continuity` would have to make as a first step `refine continuous.comp _ _`, but this required high-order unification that Lean can't cope with.

(Perhaps just leave a comment `-- The continuity tactic fails here because it would need to do higher-order unification.`)

Reply:
I've just refactored the proof and deleted the `TODO`.